File: backend/services/stigmer-server/pkg/domain/schedule/temporal/runstarter_test.go
package temporal

import (
	"context"
	"testing"
	"time"

	agentv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agent/v1"
	agentexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
	schedulev1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/schedule/v1"
	sessionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/session/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/store"
	"github.com/stigmer/stigmer/backend/libs/go/store/sqlite"
	"github.com/stretchr/testify/require"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/proto"
)

// fakeExecutionCreator records the create request and answers with a
// canned response or error.
type fakeExecutionCreator struct {
	created  *agentexecutionv1.AgentExecution
	response *agentexecutionv1.AgentExecution
	err      error
	// persistOnCreate mimics the real pipeline: the created execution
	// lands in the store, so a subsequent by-name lookup finds it.
	persistOnCreate store.Store
}

func (f *fakeExecutionCreator) Create(ctx context.Context, execution *agentexecutionv1.AgentExecution) (*agentexecutionv1.AgentExecution, error) {
	f.created = execution
	if f.err != nil {
		return nil, f.err
	}
	response := f.response
	if response == nil {
		// proto.Clone, never a struct copy: proto messages carry an
		// internal mutex, so copying trips vet's copylocks.
		clone := proto.Clone(execution).(*agentexecutionv1.AgentExecution)
		clone.Metadata = &apiresource.ApiResourceMetadata{
			Id:   "aex_01CREATED",
			Org:  execution.GetMetadata().GetOrg(),
			Name: execution.GetMetadata().GetName(),
			Slug: execution.GetMetadata().GetName(),
		}
		response = clone
	}
	if f.persistOnCreate != nil {
		if err := f.persistOnCreate.SaveResource(ctx,
			apiresourcekind.ApiResourceKind_agent_execution,
			response.GetMetadata().GetId(), response); err != nil {
			return nil, err
		}
	}
	return response, nil
}

func newStarterFixture(t *testing.T) (store.Store, *fakeExecutionCreator, *RunStarter) {
	t.Helper()
	st, err := sqlite.NewStore(t.TempDir() + "/test.sqlite")
	require.NoError(t, err)
	t.Cleanup(func() { st.Close() })
	creator := &fakeExecutionCreator{persistOnCreate: st}
	return st, creator, NewRunStarter(st, LoadConfig(), creator)
}

func seedAgent(t *testing.T, st store.Store) *agentv1.Agent {
	t.Helper()
	agent := &agentv1.Agent{
		Metadata: &apiresource.ApiResourceMetadata{
			Id: "agt_01TARGET", Org: "acme", Slug: "fee-bot", Name: "fee-bot"},
		Spec: &agentv1.AgentSpec{Instructions: "You send fee reminders."},
	}
	require.NoError(t, st.SaveResource(context.Background(),
		apiresourcekind.ApiResourceKind_agent, agent.GetMetadata().GetId(), agent))
	return agent
}

var starterFireTime = time.Date(2026, 8, 3, 9, 0, 0, 0, time.UTC)

func TestScheduledExecutionName_PinnedByteForByte(t *testing.T) {
	// THE idempotency key, byte-identical to the cloud's
	// scheduledExecutionName (its test pins the same example) — two
	// editions must never name the same fire's run differently. Every
	// character is already slug-shaped, so the name survives the create
	// pipeline's slug generator unchanged.
	require.Equal(t, "sch-01test-20260803t090000z",
		ScheduledExecutionName("sch_01TEST", starterFireTime))
	require.Equal(t, "sch-01test-20260803t090000z",
		ScheduledExecutionName("sch_01TEST", starterFireTime.Add(400*time.Millisecond)),
		"sub-second fire times truncate — the whole-second granularity of the workflow-id suffix")
	require.NotEqual(t,
		ScheduledExecutionName("sch_01TEST", starterFireTime),
		ScheduledExecutionName("sch_01TEST", starterFireTime.Add(time.Second)),
		"two fires of one schedule never collide")
}

func TestComposeMessage_FireContextLineIsContract(t *testing.T) {
	st, _, _ := newStarterFixture(t)
	_ = st
	schedule := seedSchedule2(t, nil)

	// Byte-identical to the cloud's composeMessage output: the runner
	// injects no current date into any prompt, so this line is the only
	// way the model knows "today" — and both editions must say it the
	// same way. 09:00 UTC = 14:30 IST.
	require.Equal(t,
		"Send fee reminders.\n\n(Scheduled fire time: Monday, 2026-08-03 14:30 (Asia/Kolkata))",
		ComposeMessage(schedule, starterFireTime))
}

func TestComposeMessage_UnknownZoneDegradesToUTC(t *testing.T) {
	schedule := seedSchedule2(t, func(s *schedulev1.Schedule) { s.Spec.TimeZone = "Not/AZone" })
	require.Equal(t,
		"Send fee reminders.\n\n(Scheduled fire time: Monday, 2026-08-03 09:00 (UTC))",
		ComposeMessage(schedule, starterFireTime),
		"a slightly wrong-timezone reminder beats a dead fire")
}

// seedSchedule2 builds (without persisting) the starter tests' schedule.
func seedSchedule2(t *testing.T, mutate func(*schedulev1.Schedule)) *schedulev1.Schedule {
	t.Helper()
	schedule := &schedulev1.Schedule{
		Metadata: &apiresource.ApiResourceMetadata{
			Id: "sch_01TEST", Org: "acme", Slug: "fee-reminders", Name: "fee-reminders"},
		Spec: &schedulev1.ScheduleSpec{
			Cron: "0 9 * * *", TimeZone: "Asia/Kolkata", Enabled: true,
			Target: &schedulev1.ScheduleSpec_Agent{Agent: &agentexecutionv1.AgentInvocation{
				AgentRef: &apiresource.ApiResourceReference{
					Kind: apiresourcekind.ApiResourceKind_agent, Org: "acme", Slug: "fee-bot"},
				Message: "Send fee reminders.",
			}},
		},
		Status: &schedulev1.ScheduleStatus{},
	}
	if mutate != nil {
		mutate(schedule)
	}
	return schedule
}

func persistSchedule(t *testing.T, st store.Store, schedule *schedulev1.Schedule) {
	t.Helper()
	require.NoError(t, st.SaveResource(context.Background(),
		apiresourcekind.ApiResourceKind_schedule, schedule.GetMetadata().GetId(), schedule))
}

func TestStartRun_ShapesTheUnattendedRun(t *testing.T) {
	st, creator, starter := newStarterFixture(t)
	seedAgent(t, st)
	schedule := seedSchedule2(t, nil)
	persistSchedule(t, st, schedule)

	outcome, err := starter.StartRun(context.Background(), schedule, starterFireTime)

	require.NoError(t, err)
	started, ok := outcome.(RunStartedOutcome)
	require.True(t, ok, "expected RunStartedOutcome, got %T", outcome)
	require.Equal(t, "aex_01CREATED", started.ExecutionID)
	require.False(t, started.AlreadyExisted)

	request := creator.created
	require.Equal(t, "sch-01test-20260803t090000z", request.GetMetadata().GetName(),
		"the deterministic name IS the request name")
	require.Equal(t, "acme", request.GetMetadata().GetOrg(),
		"OSS stamps the schedule's org directly — no token scope step exists here")
	require.Equal(t, "agt_01TARGET", request.GetSpec().GetAgentId())
	require.Contains(t, request.GetSpec().GetMessage(), "(Scheduled fire time: ")
	require.Equal(t, "Scheduled run: fee-reminders", request.GetSpec().GetSessionSpec().GetSubject(),
		"the pinned subject — which also opts the session out of LLM titling")
	require.Equal(t, agentexecutionv1.ApprovalMode_APPROVAL_MODE_UNATTENDED,
		request.GetSpec().GetExecutionConfig().GetApprovalMode(),
		"a gated tool with no approver would park the run forever — unattended is correctness")
	require.EqualValues(t, 20, request.GetSpec().GetExecutionConfig().GetMaxToolRounds())
	require.InDelta(t, 1.00, request.GetSpec().GetExecutionConfig().GetMaxCostUsd(), 0.001)
	require.Equal(t, agentexecutionv1.ServiceTier_SERVICE_TIER_UNSPECIFIED,
		request.GetSpec().GetExecutionConfig().GetServiceTier(),
		"no owner tier means unset — the runner resolves STANDARD, never the provider default")

	// The invocation's session half defaults to the platform: no owner
	// harness means unset (native applies), no workspace means none.
	require.Equal(t, sessionv1.Harness_HARNESS_UNSPECIFIED,
		request.GetSpec().GetSessionSpec().GetHarness())
	require.Empty(t, request.GetSpec().GetSessionSpec().GetWorkspaceEntries())

	// The run pointer landed on status.
	stored := &schedulev1.Schedule{}
	require.NoError(t, st.GetResource(context.Background(),
		apiresourcekind.ApiResourceKind_schedule, schedule.GetMetadata().GetId(), stored))
	require.Equal(t, "aex_01CREATED", stored.GetStatus().GetLastExecutionId())
}

func TestStartRun_CarriesTheInvocationSessionShape(t *testing.T) {
	// DD-018 D-3: the owner's harness, workspace, and run_config travel
	// from the invocation onto the fresh per-fire session and its
	// execution profile — model replaces the platform value outright,
	// bounds clamp min(owner, platform).
	st, creator, starter := newStarterFixture(t)
	seedAgent(t, st)
	schedule := seedSchedule2(t, func(s *schedulev1.Schedule) {
		invocation := s.GetSpec().GetAgent()
		invocation.Harness = sessionv1.Harness_HARNESS_CURSOR
		invocation.WorkspaceEntries = []*sessionv1.WorkspaceEntry{{
			Name: "docs",
			Source: &sessionv1.WorkspaceSource{Source: &sessionv1.WorkspaceSource_GitRepo{
				GitRepo: &sessionv1.GitRepoSource{Url: "https://github.com/acme/docs.git"},
			}},
		}}
		invocation.RunConfig = &agentexecutionv1.RunConfig{
			ModelName: "claude-sonnet-4-6",
			// Above the platform ceiling (1.00 / 20 in the fixture
			// config): the platform cap must win.
			MaxCostUsd:    5.00,
			MaxToolRounds: 100,
			ServiceTier:   agentexecutionv1.ServiceTier_SERVICE_TIER_FAST,
		}
	})
	persistSchedule(t, st, schedule)

	_, err := starter.StartRun(context.Background(), schedule, starterFireTime)
	require.NoError(t, err)

	sessionSpec := creator.created.GetSpec().GetSessionSpec()
	require.Equal(t, sessionv1.Harness_HARNESS_CURSOR, sessionSpec.GetHarness(),
		"the owner's harness choice reaches the session")
	require.Len(t, sessionSpec.GetWorkspaceEntries(), 1)
	require.Equal(t, "https://github.com/acme/docs.git",
		sessionSpec.GetWorkspaceEntries()[0].GetSource().GetGitRepo().GetUrl())

	config := creator.created.GetSpec().GetExecutionConfig()
	require.Equal(t, "claude-sonnet-4-6", config.GetModelName(),
		"model replaces the platform value outright — it is the owner's spend")
	require.InDelta(t, 1.00, config.GetMaxCostUsd(), 0.001,
		"the owner can lower spend, never raise it past the platform")
	require.EqualValues(t, 20, config.GetMaxToolRounds())
	require.Equal(t, agentexecutionv1.ServiceTier_SERVICE_TIER_FAST, config.GetServiceTier(),
		"the owner's stored tier is honored, not silently dropped (#357 write-time validation owns coherence)")
}

func TestStartRun_TheClockOwnsItsIdempotency(t *testing.T) {
	// DD-015 D-F: the OSS create pipeline deliberately has no duplicate
	// check, so a retried fire finds its own prior execution by the
	// deterministic name BEFORE creating — one reminder per fire, by
	// construction.
	st, creator, starter := newStarterFixture(t)
	seedAgent(t, st)
	schedule := seedSchedule2(t, nil)
	persistSchedule(t, st, schedule)

	first, err := starter.StartRun(context.Background(), schedule, starterFireTime)
	require.NoError(t, err)
	require.False(t, first.(RunStartedOutcome).AlreadyExisted)

	creator.created = nil
	second, err := starter.StartRun(context.Background(), schedule, starterFireTime)
	require.NoError(t, err)
	require.True(t, second.(RunStartedOutcome).AlreadyExisted, "the retry finds the winner")
	require.Equal(t, first.(RunStartedOutcome).ExecutionID, second.(RunStartedOutcome).ExecutionID)
	require.Nil(t, creator.created, "the retry must NOT create a second execution — that is the double reminder")
}

func TestStartRun_DanglingAgentIsTheDeterministicFailure(t *testing.T) {
	st, creator, starter := newStarterFixture(t)
	// No agent seeded: the reference dangles (deleted target — no
	// cascade by contract).
	schedule := seedSchedule2(t, nil)
	persistSchedule(t, st, schedule)

	outcome, err := starter.StartRun(context.Background(), schedule, starterFireTime)

	require.NoError(t, err)
	missing, ok := outcome.(RunTargetMissingOutcome)
	require.True(t, ok, "expected RunTargetMissingOutcome, got %T", outcome)
	require.Equal(t, "target agent acme/fee-bot not found", missing.Reason,
		"the start-failure copy is cross-edition contract — the pause reason builds on it")
	require.Nil(t, creator.created, "no execution may exist for a fire that could not resolve its target")
}

func TestStartRun_DeterministicRefusalsBecomeOutcomes(t *testing.T) {
	// A gate's refusal is a verdict for the streak; an infrastructure
	// failure is a retry. Conflating them either hides real failures or
	// pauses schedules on network blips.
	st, creator, starter := newStarterFixture(t)
	seedAgent(t, st)
	schedule := seedSchedule2(t, nil)
	persistSchedule(t, st, schedule)

	creator.err = status.Error(codes.FailedPrecondition, "the launch gate's own words")
	outcome, err := starter.StartRun(context.Background(), schedule, starterFireTime)
	require.NoError(t, err)
	refused, ok := outcome.(RunRefusedOutcome)
	require.True(t, ok, "expected RunRefusedOutcome, got %T", outcome)
	require.Equal(t, "the launch gate's own words", refused.Reason)

	creator.err = status.Error(codes.Unavailable, "engine down")
	_, err = starter.StartRun(context.Background(), schedule, starterFireTime)
	require.Error(t, err, "infrastructure failures propagate so the activity retries")
}

func TestStartRun_ModelPinningBackstop(t *testing.T) {
	// stigmer/stigmer#362: rows written before the write-time rule (an
	// explicit-cursor schedule with no pinned model) must refuse at fire
	// time — a deterministic verdict for the failure streak — instead of
	// running Auto at the provider account default's price.
	st, creator, starter := newStarterFixture(t)
	seedAgent(t, st)

	unpinned := seedSchedule2(t, func(s *schedulev1.Schedule) {
		s.Spec.GetAgent().Harness = sessionv1.Harness_HARNESS_CURSOR
	})
	persistSchedule(t, st, unpinned)

	outcome, err := starter.StartRun(context.Background(), unpinned, starterFireTime)
	require.NoError(t, err)
	refused, ok := outcome.(RunRefusedOutcome)
	require.True(t, ok, "expected RunRefusedOutcome, got %T", outcome)
	require.Equal(t,
		"spec.agent.run_config.model_name must name a pinned model when the run would use "+
			"the Cursor harness — with no pinned model Cursor runs Auto, whose price variant "+
			"follows the provider account's out-of-band default speed setting (stigmer/stigmer#362)",
		refused.Reason,
		"the copy matches the write-time rule — the fix is the same either way")
	require.Nil(t, creator.created, "no execution may exist for a refused fire")

	// A pinned model fires; the native default (unset harness) is exempt.
	pinned := seedSchedule2(t, func(s *schedulev1.Schedule) {
		s.Spec.GetAgent().Harness = sessionv1.Harness_HARNESS_CURSOR
		s.Spec.GetAgent().RunConfig = &agentexecutionv1.RunConfig{ModelName: "composer-2.5"}
	})
	persistSchedule(t, st, pinned)
	outcome, err = starter.StartRun(context.Background(), pinned, starterFireTime)
	require.NoError(t, err)
	require.IsType(t, RunStartedOutcome{}, outcome, "a pinned cursor schedule fires")

	native := seedSchedule2(t, func(s *schedulev1.Schedule) {
		s.Metadata.Id = "sch_01NATIVE"
	})
	persistSchedule(t, st, native)
	outcome, err = starter.StartRun(context.Background(), native, starterFireTime)
	require.NoError(t, err)
	require.IsType(t, RunStartedOutcome{}, outcome,
		"the native default is exempt — an empty model is deterministic there")
}

File: backend/services/stigmer-server/src/temporal/schedule/__tests__/run-starter.test.ts
/**
 * Pins the RunStarter against Go's runstarter_test.go: the deterministic
 * execution name (THE idempotency key, byte-pinned on both editions), the
 * fire-context message rendering (the model's only "today"), the
 * owner-vs-platform clamps, the find-before-create idempotency, the
 * refusal-code partition, and the execution-request shape.
 */
import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

import { Code, ConnectError } from "@connectrpc/connect";
import { create } from "@bufbuild/protobuf";
import { afterAll, beforeAll, describe, expect, it } from "vitest";

import { AgentSchema } from "@stigmer/protos/ai/stigmer/agentic/agent/v1/api_pb";
import { AgentExecutionSchema } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/api_pb";
import type { AgentExecution } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/api_pb";
import {
  ApprovalMode,
  ServiceTier,
} from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/enum_pb";
import { ScheduleSchema } from "@stigmer/protos/ai/stigmer/agentic/schedule/v1/api_pb";
import type { Schedule } from "@stigmer/protos/ai/stigmer/agentic/schedule/v1/api_pb";
import { Harness } from "@stigmer/protos/ai/stigmer/agentic/session/v1/enum_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { createLogger } from "../../../boot/logger.js";
import { SqliteStore } from "../../../store/sqlite/store.js";
import { ScheduleTemporalConfig } from "../config.js";
import {
  RunStarter,
  SCHEDULE_ID_LABEL_KEY,
  SESSION_SUBJECT_PREFIX,
  clampedRunBound,
  composeMessage,
  scheduledExecutionName,
} from "../run-starter.js";

const silentLogger = createLogger({ level: "error", pretty: false, write: () => {} });

const NOMINAL = new Date("2026-08-25T09:30:00Z");

describe("scheduledExecutionName — THE idempotency key (byte-pinned)", () => {
  it("lowercases the id, maps underscores to hyphens, appends the 20060102t150405z time", () => {
    expect(scheduledExecutionName("SCH_01ABC", NOMINAL)).toBe(
      "sch-01abc-20260825t093000z",
    );
  });

  it("truncates the nominal time to whole seconds", () => {
    expect(scheduledExecutionName("sch_x", new Date("2026-08-25T09:30:00.999Z"))).toBe(
      "sch-x-20260825t093000z",
    );
  });

  it("every character is slug-shaped (the name IS the execution's slug)", () => {
    expect(scheduledExecutionName("sch_01ABC", NOMINAL)).toMatch(/^[a-z0-9-]+$/);
  });
});

function scheduleWith(overrides?: {
  message?: string;
  timeZone?: string;
  harness?: Harness;
  modelName?: string;
  maxToolRounds?: number;
  maxCostUsd?: number;
  serviceTier?: ServiceTier;
  agentSlug?: string;
  slug?: string;
}): Schedule {
  return create(ScheduleSchema, {
    metadata: { id: "sch_01test", org: "acme", slug: overrides?.slug ?? "daily" },
    spec: {
      cron: "0 9 * * *",
      timeZone: overrides?.timeZone ?? "Asia/Kolkata",
      enabled: true,
      target: {
        case: "agent",
        value: {
          agentRef: { slug: overrides?.agentSlug ?? "helper" },
          message: overrides?.message ?? "Do the morning rounds",
          harness: overrides?.harness ?? Harness.UNSPECIFIED,
          runConfig: {
            modelName: overrides?.modelName ?? "",
            maxToolRounds: overrides?.maxToolRounds ?? 0,
            maxCostUsd: overrides?.maxCostUsd ?? 0,
            serviceTier: overrides?.serviceTier ?? ServiceTier.UNSPECIFIED,
          },
        },
      },
    },
  });
}

describe("composeMessage — the fire-context line (byte contract)", () => {
  it("renders 'Monday, 2006-01-02 15:04' in the schedule's zone with the zone-name echo", () => {
    // 2026-08-25T09:30:00Z is 15:00 IST on Tuesday 2026-08-25.
    expect(composeMessage(scheduleWith(), NOMINAL)).toBe(
      "Do the morning rounds\n\n(Scheduled fire time: Tuesday, 2026-08-25 15:00 (Asia/Kolkata))",
    );
  });

  it("an empty zone renders in UTC and echoes 'UTC' (Go LoadLocation(''))", () => {
    expect(composeMessage(scheduleWith({ timeZone: "" }), NOMINAL)).toBe(
      "Do the morning rounds\n\n(Scheduled fire time: Tuesday, 2026-08-25 09:30 (UTC))",
    );
  });

  it("an unloadable zone degrades to UTC (a slightly wrong-timezone reminder beats a dead fire)", () => {
    expect(composeMessage(scheduleWith({ timeZone: "Mars/Olympus" }), NOMINAL)).toBe(
      "Do the morning rounds\n\n(Scheduled fire time: Tuesday, 2026-08-25 09:30 (UTC))",
    );
  });

  it("uses h23 hours (Go's 15 layout: 00-23, never 24)", () => {
    const midnight = new Date("2026-08-25T00:05:00Z");
    expect(composeMessage(scheduleWith({ timeZone: "UTC" }), midnight)).toContain(
      "Tuesday, 2026-08-25 00:05 (UTC)",
    );
  });
});

describe("clampedRunBound — min(owner, platform), zero = unset", () => {
  it.each([
    [0, 20, 20], // owner unset → platform
    [5, 20, 5], // both set → lower wins
    [50, 20, 20], // owner cannot raise past the platform
    [5, 0, 5], // platform unset → owner stands
    [0, 0, 0], // both unset
    [-1, 20, 20], // negative reads as unset
  ])("owner=%d platform=%d → %d", (owner, platform, want) => {
    expect(clampedRunBound(owner, platform)).toBe(want);
  });
});

describe("RunStarter.startRun — against a real store", () => {
  let dir: string;
  let store: SqliteStore;

  const config = new ScheduleTemporalConfig(
    "schedule_stigmer",
    60,
    24,
    5,
    60,
    true,
    5,
    20,
    1.0,
    90,
  );

  beforeAll(async () => {
    dir = mkdtempSync(path.join(tmpdir(), "run-starter-test-"));
    store = SqliteStore.open(path.join(dir, "test.db"));
    await store.saveResource(
      ApiResourceKind.agent,
      "agt_01helper",
      AgentSchema,
      create(AgentSchema, {
        metadata: { id: "agt_01helper", org: "acme", slug: "helper" },
      }),
    );
    await store.saveResource(
      ApiResourceKind.schedule,
      "sch_01test",
      ScheduleSchema,
      scheduleWith(),
    );
  });

  afterAll(async () => {
    await store.close();
    rmSync(dir, { recursive: true, force: true });
  });

  function starter(create_: (execution: AgentExecution) => Promise<AgentExecution>) {
    return new RunStarter({
      store,
      config,
      executions: { create: create_ },
      logger: silentLogger,
    });
  }

  it("shapes the execution request: name, org, label, UNATTENDED, subject prefix, fire-context message", async () => {
    let seen: AgentExecution | undefined;
    const runStarter = starter(async (execution) => {
      seen = execution;
      return create(AgentExecutionSchema, {
        metadata: { id: "aex_01new", org: "acme" },
      });
    });

    const outcome = await runStarter.startRun(scheduleWith(), NOMINAL);
    expect(outcome).toEqual({
      kind: "started",
      executionId: "aex_01new",
      alreadyExisted: false,
    });

    expect(seen?.apiVersion).toBe("agentic.stigmer.ai/v1");
    expect(seen?.kind).toBe("AgentExecution");
    expect(seen?.metadata?.name).toBe("sch-01test-20260825t093000z");
    expect(seen?.metadata?.org).toBe("acme");
    expect(seen?.metadata?.labels[SCHEDULE_ID_LABEL_KEY]).toBe("sch_01test");
    expect(seen?.spec?.agentId).toBe("agt_01helper");
    expect(seen?.spec?.message).toContain("(Scheduled fire time: ");
    expect(seen?.spec?.sessionSpec?.subject).toBe(`${SESSION_SUBJECT_PREFIX}daily`);
    expect(seen?.spec?.executionConfig?.approvalMode).toBe(ApprovalMode.UNATTENDED);
    // Platform profile applies when the owner set nothing.
    expect(seen?.spec?.executionConfig?.maxToolRounds).toBe(20);
    expect(seen?.spec?.executionConfig?.maxCostUsd).toBe(1.0);
    // Unset tier/model stay unset (never stamped).
    expect(seen?.spec?.executionConfig?.modelName).toBe("");
    expect(seen?.spec?.executionConfig?.serviceTier).toBe(ServiceTier.UNSPECIFIED);

    // Success stamped last_execution_id on the row.
    const row = await store.getResource(
      ApiResourceKind.schedule,
      "sch_01test",
      ScheduleSchema,
    );
    expect(row.status?.lastExecutionId).toBe("aex_01new");
  });

  it("finds an existing execution by the deterministic name instead of creating (idempotent retry)", async () => {
    await store.saveResource(
      ApiResourceKind.agent_execution,
      "aex_01prior",
      AgentExecutionSchema,
      create(AgentExecutionSchema, {
        metadata: {
          id: "aex_01prior",
          org: "acme",
          slug: scheduledExecutionName("sch_01test", NOMINAL),
        },
      }),
    );
    const runStarter = starter(async () => {
      throw new Error("create must not be called on the idempotent path");
    });
    const outcome = await runStarter.startRun(scheduleWith(), NOMINAL);
    expect(outcome).toEqual({
      kind: "started",
      executionId: "aex_01prior",
      alreadyExisted: true,
    });
    await store.deleteResource(ApiResourceKind.agent_execution, "aex_01prior");
  });

  // The scheduleFireCaller seam (stigmer-cloud#572): composed → every
  // create carries the per-fire minted caller; absent → the create
  // carries none (the internal lane, byte-identical); mint failure → an
  // infrastructure throw the tick retries; the idempotent path never
  // mints (a found winner needs no credential).
  describe("fire-caller mint (stigmer-cloud#572)", () => {
    const mintedCaller = {
      identityId: "ida_schedule_acct",
      callerClass: "schedule",
      issuer: "stigmer",
      rawToken: "minted.jwt",
    };

    it("propagates the minted caller to the create, minting per (org, scheduleId)", async () => {
      const mintCalls: Array<{ org: string; scheduleId: string }> = [];
      let seenCaller: unknown = "unset";
      const runStarter = new RunStarter({
        store,
        config,
        executions: {
          create: async (_execution, fireCaller) => {
            seenCaller = fireCaller;
            return create(AgentExecutionSchema, {
              metadata: { id: "aex_01minted", org: "acme" },
            });
          },
        },
        fireCallerMint: {
          mintFireCaller: async (org, scheduleId) => {
            mintCalls.push({ org, scheduleId });
            return mintedCaller;
          },
        },
        logger: silentLogger,
      });

      const outcome = await runStarter.startRun(scheduleWith(), NOMINAL);
      expect(outcome).toEqual({
        kind: "started",
        executionId: "aex_01minted",
        alreadyExisted: false,
      });
      expect(mintCalls).toEqual([{ org: "acme", scheduleId: "sch_01test" }]);
      expect(seenCaller).toBe(mintedCaller);
    });

    it("passes no caller when no mint is composed (the OSS internal lane)", async () => {
      let seenCaller: unknown = "unset";
      const runStarter = new RunStarter({
        store,
        config,
        executions: {
          create: async (_execution, fireCaller) => {
            seenCaller = fireCaller;
            return create(AgentExecutionSchema, {
              metadata: { id: "aex_01plain", org: "acme" },
            });
          },
        },
        logger: silentLogger,
      });
      const outcome = await runStarter.startRun(scheduleWith(), NOMINAL);
      expect(outcome.kind).toBe("started");
      expect(seenCaller).toBeUndefined();
    });

    it("throws (retryable) when the mint fails — never a silent internal fallback", async () => {
      const runStarter = new RunStarter({
        store,
        config,
        executions: {
          create: async () => {
            throw new Error("create must not run when the mint failed");
          },
        },
        fireCallerMint: {
          mintFireCaller: async () => {
            throw new Error("account provisioning unavailable");
          },
        },
        logger: silentLogger,
      });
      await expect(runStarter.startRun(scheduleWith(), NOMINAL)).rejects.toThrow(
        /mint schedule fire caller for sch_01test: account provisioning unavailable/,
      );
    });

    it("never mints on the idempotent path (the winner needs no credential)", async () => {
      await store.saveResource(
        ApiResourceKind.agent_execution,
        "aex_01winner",
        AgentExecutionSchema,
        create(AgentExecutionSchema, {
          metadata: {
            id: "aex_01winner",
            org: "acme",
            slug: scheduledExecutionName("sch_01test", NOMINAL),
          },
        }),
      );
      const runStarter = new RunStarter({
        store,
        config,
        executions: {
          create: async () => {
            throw new Error("create must not run on the idempotent path");
          },
        },
        fireCallerMint: {
          mintFireCaller: async () => {
            throw new Error("mint must not run on the idempotent path");
          },
        },
        logger: silentLogger,
      });
      const outcome = await runStarter.startRun(scheduleWith(), NOMINAL);
      expect(outcome).toEqual({
        kind: "started",
        executionId: "aex_01winner",
        alreadyExisted: true,
      });
      await store.deleteResource(ApiResourceKind.agent_execution, "aex_01winner");
    });
  });

  it("answers targetMissing with the byte-pinned copy when the agent is gone", async () => {
    const runStarter = starter(async () => {
      throw new Error("unreachable");
    });
    const outcome = await runStarter.startRun(
      scheduleWith({ agentSlug: "vanished" }),
      NOMINAL,
    );
    expect(outcome).toEqual({
      kind: "targetMissing",
      reason: "target agent acme/vanished not found",
    });
  });

  it("refuses via the model-pinning launch backstop for a pre-rule cursor row", async () => {
    const runStarter = starter(async () => {
      throw new Error("unreachable");
    });
    const outcome = await runStarter.startRun(
      scheduleWith({ harness: Harness.CURSOR, modelName: "" }),
      NOMINAL,
    );
    expect(outcome.kind).toBe("refused");
    if (outcome.kind === "refused") {
      expect(outcome.reason).toContain("stigmer/stigmer#362");
    }
  });

  it.each([
    Code.FailedPrecondition,
    Code.PermissionDenied,
    Code.NotFound,
    Code.InvalidArgument,
    Code.ResourceExhausted,
  ])("maps the launch-gate code %s to a refused outcome (the streak counts it)", async (code) => {
    const runStarter = starter(async () => {
      throw new ConnectError("gate refused this run", code);
    });
    const outcome = await runStarter.startRun(scheduleWith(), NOMINAL);
    expect(outcome).toEqual({ kind: "refused", reason: "gate refused this run" });
  });

  it("rethrows infrastructure codes so the activity retries", async () => {
    const runStarter = starter(async () => {
      throw new ConnectError("store exploded", Code.Internal);
    });
    await expect(runStarter.startRun(scheduleWith(), NOMINAL)).rejects.toThrow(
      "store exploded",
    );
  });

  it("re-finds the winner on AlreadyExists from the session duplicate check", async () => {
    // The pre-check misses (no row yet); create refuses AlreadyExists and
    // the winner row lands concurrently (the session auto-create race the
    // Go comment describes) — the starter must re-read the winner.
    const raceNominal = new Date("2026-08-26T09:30:00Z");
    let calls = 0;
    const runStarter = starter(async () => {
      calls++;
      await store.saveResource(
        ApiResourceKind.agent_execution,
        "aex_01winner",
        AgentExecutionSchema,
        create(AgentExecutionSchema, {
          metadata: {
            id: "aex_01winner",
            org: "acme",
            slug: scheduledExecutionName("sch_01test", raceNominal),
          },
        }),
      );
      throw new ConnectError("duplicate", Code.AlreadyExists);
    });
    const outcome = await runStarter.startRun(scheduleWith(), raceNominal);
    expect(calls).toBe(1);
    expect(outcome).toEqual({
      kind: "started",
      executionId: "aex_01winner",
      alreadyExisted: true,
    });
    await store.deleteResource(ApiResourceKind.agent_execution, "aex_01winner");
  });

  it("clamps owner run bounds by the platform profile and stamps owner-set tier/model", async () => {
    let seen: AgentExecution | undefined;
    const runStarter = starter(async (execution) => {
      seen = execution;
      return create(AgentExecutionSchema, { metadata: { id: "aex_01x", org: "acme" } });
    });
    await runStarter.startRun(
      scheduleWith({
        maxToolRounds: 50, // above the platform 20 → clamped
        maxCostUsd: 0.25, // below the platform 1.00 → stands
        modelName: " gpt-x ",
        serviceTier: ServiceTier.FAST,
      }),
      new Date("2026-08-27T09:30:00Z"),
    );
    expect(seen?.spec?.executionConfig?.maxToolRounds).toBe(20);
    expect(seen?.spec?.executionConfig?.maxCostUsd).toBe(0.25);
    expect(seen?.spec?.executionConfig?.modelName).toBe("gpt-x");
    expect(seen?.spec?.executionConfig?.serviceTier).toBe(ServiceTier.FAST);
  });
});
